cmd/health/health.go

```diff
@@ -0,0 +1,192 @@
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+package health
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"strings"
+
+	"github.com/aurae-runtime/ae/pkg/cli"
+	"github.com/aurae-runtime/ae/pkg/cli/printer"
+	"github.com/aurae-runtime/ae/pkg/client"
+	"github.com/aurae-runtime/ae/pkg/config"
+	"github.com/spf13/cobra"
+
+	aeCMD "github.com/aurae-runtime/ae/cmd"
+	discoveryv0 "github.com/aurae-runtime/ae/pkg/api/v0/discovery"
+)
+
+type outputHealthNode struct {
+	Version string `json:"version"`
+}
+
+type outputHealth struct {
+	Nodes map[string]outputHealthNode `json:"nodes"`
+}
+
+func inc(ip net.IP) {
+	for i := len(ip) - 1; i >= 0; i-- {
+		ip[i]++
+		if ip[i] > 0 {
+			break
+		}
+	}
+}
+
+func hosts(cidr string) ([]string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+
+	var ips []string
+	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); inc(ip) {
+		ips = append(ips, ip.String())
+	}
+	// remove network address and broadcast address
+	if strings.HasSuffix(ips[0], ".0") {
+		ips = ips[1:]
+	}
+	if strings.HasSuffix(ips[len(ips)-1], ".255") {
+		ips = ips[:len(ips)-1]
+	}
+	return ips, nil
+}
+
+func protocol(ip_str string) (string, error) {
+	ip := net.ParseIP(ip_str)
+	if ip == nil {
+		return "", fmt.Errorf("unable to parse IP %q", ip_str)
+	}
+	protocol := "tcp4"
+	if ip.To4() == nil {
+		protocol = "tcp6"
+	}
+	return protocol, nil
+}
+
+type option struct {
+	aeCMD.Option
+	cidr         string
+	port         uint16
+	verbose      bool
+	writer       io.Writer
+	outputFormat *cli.OutputFormat
+}
+
+func (o *option) Complete(_ []string) error {
+	return nil
+}
+
+func (o *option) Validate() error {
+	if err := o.outputFormat.Validate(); err != nil {
+		return err
+	}
+
+	if _, _, err := net.ParseCIDR(o.cidr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *option) Execute() error {
+	ctx := context.Background()
+	hosts, err := hosts(o.cidr)
+	if err != nil {
+		return err
+	}
+
+	output := &outputHealth{
+		Nodes: make(map[string]outputHealthNode),
+	}
+
+	// TODO: concurrency.
+	for _, ip_str := range hosts {
```

Comment:
Move to generic library such that we can pass an implementation of your yet-to-be-defined interface to the library.